lib/vm/src/imports.rs

```diff
@@ -17,9 +17,7 @@ pub struct Imports {
     /// space may affect Wasm runtime performance due to increased cache pressure.
     ///
     /// We make it optional so that we can free the data after use.
-    pub host_function_env_initializers: Option<
-        BoxedSlice<FunctionIndex, (Option<ImportInitializerFuncPtr>, *mut std::ffi::c_void)>,
-    >,
+    pub host_function_env_initializers: Option<BoxedSlice<FunctionIndex, ImportFunctionEnv>>,
```

Comment:
This should have the same type as `functions` field, this means that there should be no need to wrap things with an `Option`.

```suggestion
    pub host_function_env_initializers: BoxedSlice<FunctionIndex, ImportFunctionEnv>,
```

Reply:
The use of option here is for correctness / hinting at access patterns and to reduce the amount of dead memory we have: once we get this thing once it's both incorrect to access it again (in regards to initializers) and wasteful to keep it around here.

Previously, the initializers would also be destroyed after initialization making it impossible to call it multiple times.  I believe this PR changes that in order to keep some of the code simpler, but it may be worth changing it back.

Reply:
When certain data has precise semantics I find the mental model of "consumable" data to be a good one because it's more explicit about what's happening and it also reduces peak memory usage of data that is no good to anyone anymore.  The data can only be misused by keeping it around, so maybe it's better to just get rid of it.

Reply:
I might have missed, where do we consume the `host_function_env_initializers` field?

Reply:
Oh, I gues the `take` in `get_imported_function_envs` consumes it?

Reply:
If so, can we comment that on the `host_function_env_initializers`. It was not trivial for me to realize that the field was consumable!

Reply:
I agree, it may be better to create a wrapper type that's more explicit than `Option` in wasmer_types or something.

`UseOnce` or something, something that implies that it always exists at the start but once it's used, it's gone

Reply:
I think we can use `Option` for now (it will be simpler!), but just with a clear comment on why is that way :)